[asan] Make frame number checks more flexable

hau-hsu opened the pull request:
Use more flexable regex ([0-9]+) for frame number checks. Since the frame numbers might change if some functions are not inlined.

Similar to
* https://github.com/llvm/llvm-project/commit/0360f3218a13666123849f6699216bdbebe64833
* https://github.com/llvm/llvm-project/commit/404bc5ca2ae5fffa45d5240dcf52b7f134988867

hau-hsu committed:
[asan] Make frame number checks more flexable

Use more flexable regex ([0-9]+) for stack checks.
Since the stack numbers might change if some functions are inlined or not.

Similar to
* https://github.com/llvm/llvm-project/commit/0360f3218a13666123849f6699216bdbebe64833
* https://github.com/llvm/llvm-project/commit/404bc5ca2ae5fffa45d5240dcf52b7f134988867

MaskRay commented:
> Use more flexable regex ([0-9]+) for frame number checks. Since the frame numbers might change if some functions are not inlined.

Can you mention why some functions might be inlined. We use -O0 to prevent that. This change is probably still good but we need concrete examples.

hau-hsu commented:
> > Use more flexable regex ([0-9]+) for frame number checks. Since the frame numbers might change if some functions are not inlined.
> 
> Can you mention why some functions might be inlined. We use -O0 to prevent that. This change is probably still good but we need concrete examples.

Take [compiler-rt/test/asan/TestCases/Linux/stack-trace-dlclose.cpp](https://github.com/llvm/llvm-project/blob/5787f43f867d8545af2a3d5ee724964824186af4/compiler-rt/test/asan/TestCases/Linux/stack-trace-dlclose.cpp) as an example, the FileChecks are:
```
// CHECK: {{    #0 0x.* in (__interceptor_)?malloc}}
// CHECK: {{    #1 0x.* \(<unknown module>\)}}
// CHECK: {{    #2 0x.* in main}}
```
but the call stack looks like:
```
previously allocated by thread T0 here:
    #0 0x2aaeb3662a in operator[] /home/root/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:68:3
    #1 0x2aaeb3662a in GetSizeClass /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_primary32.h:203:44
    #2 0x2aaeb3662a in PointerIsMine /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_primary32.h:198:12
    #3 0x2aaeb3662a in FromPrimary /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_combined.h:121:59
    #4 0x2aaeb3662a in Allocate /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_dlsym.h:36:5
    #5 0x2aaeb3662a in malloc /home/root/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:67:12
    #6 0x3fb7319a0a  (<unknown module>)
    #7 0x2aaeb6e996 in main /home/root/llvm-project/compiler-rt/test/asan/TestCases/Linux/stack-trace-dlclose.cpp:33:13
    #8 0x3fb770dd74  (/lib/libc.so.6+0x29d74)
```

vitalybuka commented:
Is this for clang or gcc?

hau-hsu commented:
> Is this for clang or gcc?
For clang.

vitalybuka commented:
Do you have guess why this you have this issue and we don't?

vitalybuka commented:
> Do you have guess why this you have this issue and we don't?

E.g. 0360f3218a13666123849f6699216bdbebe64833 probably caused by recent Windows changes.

hau-hsu commented:
> Do you have guess why this you have this issue and we don't?

I forgot to mention, this happens on RISC-V target.
I was testing the for #66743 on RISC-V Qemu (system mode).

The stack trace in the previous comment:
```
previously allocated by thread T0 here:
    #0 0x2aaeb3662a in operator[] /home/root/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:68:3
    #1 0x2aaeb3662a in GetSizeClass /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_primary32.h:203:44
    #2 0x2aaeb3662a in PointerIsMine /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_primary32.h:198:12
    #3 0x2aaeb3662a in FromPrimary /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_combined.h:121:59
    #4 0x2aaeb3662a in Allocate /home/root/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_allocator_dlsym.h:36:5
    #5 0x2aaeb3662a in malloc /home/root/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:67:12
    #6 0x3fb7319a0a  (<unknown module>)
    #7 0x2aaeb6e996 in main /home/root/llvm-project/compiler-rt/test/asan/TestCases/Linux/stack-trace-dlclose.cpp:33:13
    #8 0x3fb770dd74  (/lib/libc.so.6+0x29d74)
```
stack 0~5 have the same PC, and are inlined functions.

I compared the executable and dwarf between RISC-V and X86, found that `Allocate` is inlined in RISC-V but not in X86:
https://github.com/llvm/llvm-project/blob/638d96843893f6981c9af8c501f6999cc7db15e6/compiler-rt/lib/asan/asan_malloc_linux.cpp#L65-L70

Thus the stack trace looks different. This happens in several test cases. So I created this PR.

hau-hsu commented:
@vitalybuka could you also help to review https://github.com/llvm/llvm-project/pull/66743?
I think there is no critical issues for the PR.
Thank you

dyung commented on compiler-rt/test/asan/TestCases/strdup_oob_test.cpp:
```diff
@@ -23,7 +23,7 @@ int main(int argc, char **argv) {
   // CHECK: #0 {{.*}}main {{.*}}strdup_oob_test.cpp:[[@LINE-2]]
   // CHECK-LABEL: allocated by thread T{{.*}} here:
-  // CHECK: #{{[01]}} {{.*}}strdup
+  // CHECK: #{{[0-9]}}+ {{.*}}strdup
```
Should the `+` be inside the double braces since I assume you are trying to use it as a regex and not explicitly match it? I believe this is causing a test failure on a bot because of this misplacement of the `+` on https://lab.llvm.org/buildbot/#/builders/77/builds/38204

hau-hsu replied:
Yeah I had the same change locally but forgot to update the commit ...
Thanks

dyung commented on compiler-rt/test/asan/TestCases/strcpy-overlap.cpp:
```diff
@@ -36,9 +36,9 @@ __attribute__((noinline)) void bad_function() {
-  // CHECK: {{#0 0x.* in .*strcpy}}
-  // CHECK: {{#1 0x.* in bad_function.*strcpy-overlap.cpp:}}[[@LINE+2]]
-  // CHECK: {{#2 0x.* in main .*strcpy-overlap.cpp:}}[[@LINE+5]]
+  // CHECK: {{#[0-9]+ 0x.* in .*strcpy .*.cpp}}
```
We are seeing a failure in this test in our internal builder due to the line looking like this:
```
#0 0x55ec8a9ad17c in strcpy.part.0 /home/jenkins/compiler-rt/lib/asan/asan_interceptors.cpp:566:5
```
Previously the match would end at `strcpy`, but now the match fails because you are expecting a space immediately following the `strcopy` which isn't the case on our internal builder.

hau-hsu replied:
That's different from my output. Thanks for fixing that :)

MaskRay commented:
> > Do you have guess why this you have this issue and we don't?
> 
> I forgot to mention, this happens on RISC-V target. I was testing the for #66743 on RISC-V Qemu (system mode).
> 

This is an important detail. Please include such details in the future when you anticipate the question "why we don't see the failure while you do..."

hau-hsu commented:
> This is an important detail. Please include such details in the future when you anticipate the question "why we don't see the failure while you do..."

Yeah sorry about that. I'll do better next time.

zmodem commented:
Chromium is seeing failures which seem related: https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8745260118721805409/+/u/package_clang/stdout

and so is green dragon: https://green.lab.llvm.org/job/llvm.org/job/clang-stage1-RA/1095/console

vitalybuka commented:
was that caused by unnecessary check`.*.cpp`  ?

hau-hsu commented:
> was that caused by unnecessary check.*.cpp ?

Yes. My bad.

I added `.*.cpp` because I got some FileCheck errors after I apply the `[0-9a-f]+` regex for frame numbers.
Then I thought they were caused by greedy matching of `*`.

For example, without `.*.cpp`, the checks are:
```
CHECK: {{    #[0-9]+ 0x.* in .*strncpy}}
CHECK: {{    #[0-9]+ 0x.* in main .*strncpy-overflow.cpp:}}[[@LINE-3]]
```
I got the error like:
```
error: CHECK: expected string not found in input
 // CHECK: {{#[0-9]+ 0x.* in bad_function.*strcpy-overlap.cpp:}}[[@LINE+2]]
```
I thought the first check (`#[0-9]+ 0x.* in .*strncpy`) matches 2 lines
```
    #0 0x2aaab49acc in strcpy .../asan_interceptors.cpp:554:5
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    #1 0x2aaab98796 in bad_function() .../strcpy-overlap.cpp:42:3
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
So I added `.*.cpp` to prevent that.
But it seems that I shouldn't do this, and this is not the root cause of the error.

I'll review the patch and send it again. Sorry about that.
